### PR TITLE
[Refactor] Decouple preprocess operation and transformation

### PR DESCRIPTION
**Pull request by lzhangzz**
Decouple preprocess operations and transformations
- operations can be reused by other modules like post-process
- transformations can be more easily maintained
- much less boilerplate code

BC breaking - redesigned transfomation interface

prerequisite: #1294

### CODE REVIEW

File: csrc/mmdeploy/operation/cuda/crop.cpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    int h = bottom - top + 1;
+    int w = right - left + 1;
+    int c = desc.shape[3];
+    auto type = desc.data_type;
+
+    TensorShape shape{1, bottom - top + 1, right - left + 1, src.desc().shape[3]};
+    TensorDesc dst_desc{device(), src.desc().data_type, shape, desc.name};
+    Tensor dst_tensor{dst_desc};
+
+    if (DataType::kINT8 == type) {
```
Comment (lvhan028) [inline csrc/mmdeploy/operation/cuda/crop.cpp:34]: How about dispatch according to `type`?

File: csrc/mmdeploy/operation/cuda/crop.cpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      if (3 == c) {
+        impl::Crop<float, 3>(input, desc.shape[2], output, h, w, top, left, cuda_stream);
+      } else if (1 == c) {
+        impl::Crop<float, 1>(input, desc.shape[2], output, h, w, top, left, cuda_stream);
+      } else {
+        MMDEPLOY_ERROR("unsupported channels {}", c);
+        return Status(eNotSupported);
+      }
+    } else {
+      MMDEPLOY_ERROR("unsupported channels {}", c);
```
Comment (lvhan028) [inline csrc/mmdeploy/operation/cuda/crop.cpp:57]: unsupported type

File: csrc/mmdeploy/preprocess/transform/center_crop.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+using namespace std;
+
+namespace mmdeploy::transform {
+
+class CenterCrop : public Transform {
+ public:
+  explicit CenterCrop(const Value& args) {
+    if (!args.contains(("crop_size"))) {
+      throw std::invalid_argument("'crop_size' is expected");
```
Comment (lvhan028) [inline csrc/mmdeploy/preprocess/transform/center_crop.cpp:19]: should we use `throw_exception` instead?

File: csrc/mmdeploy/core/model.h [hunk 4]
```diff
@@ -60,7 +60,7 @@ class MMDEPLOY_API Model {
   Result<void> Init(const void* buffer, size_t size);
 
   /**
-   * @brief Return the model's meta info
+   * @brief Return a the model's meta info
```
Comment (lvhan028) [inline csrc/mmdeploy/core/model.h:63]: ~~the~~

File: csrc/mmdeploy/utils/opencv/opencv_utils.cpp [hunk 5]
```diff
@@ -131,13 +130,52 @@ cv::Mat Crop(const cv::Mat& src, int top, int left, int bottom, int right) {
   return src(cv::Range(top, bottom + 1), cv::Range(left, right + 1)).clone();
 }
 
+template <int a0, int a1, int a2, typename T>
```
Comment (lvhan028) [inline csrc/mmdeploy/utils/opencv/opencv_utils.cpp:133]: how about use `C0, C1, C2` instead

File: csrc/mmdeploy/preprocess/transform/transform.h [hunk 6]
```diff
@@ -6,79 +6,35 @@
...
+  MMDEPLOY_REGISTER_FACTORY_FUNC(::mmdeploy::transform::Transform, desc, [](const Value& config) { \
+    return std::make_unique<type>(config);                                                         \
+  });
 
- protected:
-  std::string specified_platform_;
-  std::string runtime_platform_;
-  std::vector<std::string> candidate_platforms_;
-};
+#define MMDEPLOY_REGISTER_TRANSFORM(type) MMDEPLOY_REGISTER_TRANSFORM2(type, (type, 0))
```
Comment (lvhan028) [inline csrc/mmdeploy/preprocess/transform/transform.h:32]: Why not put `MMDEPLOY_REISTER_TRANSFORM` and `MMDEPLOY_REGISTER_TRANSFORM2` together?

File: csrc/mmdeploy/operation/cuda/crop.cpp (hunk 1 above)
Comment (lzhangzz) [reply csrc/mmdeploy/operation/cuda/crop.cpp:34]: It's just 4 branches. I welcome anyone who is not me to refactor it.

File: csrc/mmdeploy/codebase/mmaction/cuda/format_shape_impl.cpp [hunk 7]
```diff
@@ -16,27 +15,28 @@ namespace cuda {
     }                                                                          \
   } while (0);
 
-class FormatShapeImpl : public ::mmdeploy::FormatShapeImpl {
+class FormatShapeImpl : public FormatShapeOp {
  public:
-  explicit FormatShapeImpl(const Value& args) : ::mmdeploy::FormatShapeImpl(args) {
+  explicit FormatShapeImpl(std::string input_format) : FormatShapeOp(std::move(input_format)) {
```
Comment (lvhan028) [inline csrc/mmdeploy/codebase/mmaction/cuda/format_shape_impl.cpp:20]: May use reference

File: csrc/mmdeploy/codebase/mmaction/format_shape.cpp [hunk 8]
```diff
@@ -2,81 +2,62 @@
...
+Result<void> FormatShape::Apply(Value& data) {
+  MMDEPLOY_DEBUG("input: {}", data);
 
-  if (!input.is_array()) {
+  if (!data.is_array()) {
     MMDEPLOY_ERROR("input of format shape should be array");
     return Status(eInvalidArgument);
   }
-  if (!(input[0].contains("img") || input[0].contains("img"))) {
+  if (!(data[0].contains("img") || data[0].contains("img"))) {
```
Comment (lvhan028) [inline csrc/mmdeploy/codebase/mmaction/format_shape.cpp:26]: duplicate condition

Comment (irexyc) [reply csrc/mmdeploy/codebase/mmaction/format_shape.cpp:26]: one should be imgs

> duplicate condition. Maybe one of them is `imgs`?

File: csrc/mmdeploy/utils/opencv/opencv_utils.cpp [hunk 9]
```diff
@@ -169,88 +207,67 @@ cv::Mat Transpose(const cv::Mat& src) {
...
-        return {};
+namespace {
+
+class ColorConversionTable {
+  static constexpr auto kSize = static_cast<size_t>(PixelFormat::kCOUNT);
+
+  int codes_[kSize][kSize]{};
+
+  // until we have "Deducing `this`" in C++23
+  template <typename Self>
```
Comment (lvhan028) [inline csrc/mmdeploy/utils/opencv/opencv_utils.cpp:218]: Why not `get_impl(src, dst)`

**Review by lvhan028 — APPROVED** [review]

**Review by irexyc — APPROVED** [review]

**Pull request merged**